File: LeetCode/3562. Maximum Profit from Trading Stocks with Discounts/solution.cpp
class Solution {
public:
    vector<vector<int>> g;
    vector<int> parent;

    vector<vector<vector<int>>> memo1;
    vector<vector<vector<int>>> memo2;

    int dfs1(int cur, int flag, int b, const vector<int>& present, const vector<int>& future) {
        if (memo1[cur][flag][b] != -1) {
            return memo1[cur][flag][b];
        }
        int cost = (flag == 1) ? (present[cur] / 2) : present[cur];
        int ans = dfs2(cur, 0, 0, b, present, future);
        if (cost <= b) {
            ans = max(ans, dfs2(cur, 0, 1, b - cost, present, future) + future[cur] - cost);
        }
        return memo1[cur][flag][b] = ans;
    }

    int dfs2(int pre, int i, int flag, int b, const vector<int>& present, const vector<int>& future) {
        
        int m = g[pre].size();
        if (i == m) {
            return 0;
        }
        int cur = g[pre][i];
        if (memo2[cur][flag][b] != -1) {
            return memo2[cur][flag][b];
        }
        int ans = 0;
        for (int nb = 0; nb <= b; ++nb) {
            ans = max(ans, dfs1(cur, flag, nb, present, future) + dfs2(pre, i + 1, flag, b - nb, present, future));
        }
        return memo2[cur][flag][b] = ans;
    }

    int maxProfit(int n, vector<int>& present, vector<int>& future, vector<vector<int>>& hierarchy, int budget) {
        g.resize(n);
        parent.resize(n);

        for (const auto& edge : hierarchy) {
            int u = edge[0] - 1;
            int v = edge[1] - 1;
            g[u].push_back(v);
            parent[v] = u;
        }

        memo1.assign(n, vector<vector<int>>(2, vector<int>(budget + 1, -1)));
        memo2.assign(n, vector<vector<int>>(2, vector<int>(budget + 1, -1)));

        int res = dfs1(0, 0, budget, present, future);

        return res;
    }
};

File: LeetCode/3562. Maximum Profit from Trading Stocks with Discounts/solution.py
class Solution:
    def maxProfit(self, n: int, present: List[int], future: List[int], hierarchy: List[List[int]], budget: int) -> int:
        g = defaultdict(list)
        parent = [0]*n  
        for u, v in hierarchy:
            u -= 1
            v -= 1
            g[u].append(v)
            parent[v] = u

        @lru_cache(None)
        def dfs1(cur,flag,b):
            cost = present[cur]//2 if flag else present[cur]
            ans = dfs2(cur,0,0,b)
            if cost <= b:
                ans = max(ans,dfs2(cur,0,1,b-cost)+future[cur]-cost)
            return ans 

        @lru_cache(None)
        def dfs2(cur,i,flag,b):
            m = len(g[cur])
            if i == m: return 0 
            nxt = g[cur][i]
            if nxt == parent[cur]: return dfs2(cur,i+1,flag,b)
            ans = 0
            for nb in range(b+1):
                ans = max(ans,dfs1(nxt,flag,nb)+dfs2(cur,i+1,flag,b-nb))
            return ans 

        res = dfs1(0,0,budget)
        dfs1.cache_clear()
        dfs2.cache_clear()
        return res 

